packages/core/src/common/message-rpc/rpc-message-encoder.ts

```diff
@@ -119,11 +122,12 @@ export interface ValueEncoder {
      * Write the given value to the buffer. Will only be called if {@link is(value)} returns true.
      * @param buf The buffer to write to
      * @param value The value to be written
+     * @param visitedObjects The collection of already visited (i.e. encoded) objects. Used to detect circular references
      * @param recursiveEncode A function that will use the encoders registered on the {@link MessageEncoder}
      * to write a value to the underlying buffer. This is used mostly to write structures like an array
      * without having to know how to encode the values in the array
      */
-    write(buf: WriteBuffer, value: any, recursiveEncode?: (buf: WriteBuffer, value: any) => void): void;
+    write(buf: WriteBuffer, value: any, visitedObjects: WeakSet<object>, recursiveEncode?: (buf: WriteBuffer, value: any, visitedObjects: WeakSet<object>) => void): void;
```

Comment:
Doesn't a regular set work just as well? We're never holding on to a reference to it anyway, so retention is not a problem.

Reply:
It would work as well. However, ciruclar reference dection is the prime usecase for weak sets and its even mentioned in the [doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakSet#use_case_detecting_circular_references)

>The number of objects or their traversal order is immaterial, so a WeakSet is more suitable (and performant) than a [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) for tracking object references, especially if a very large number of objects is involved.

So this is mainly a performance optimization. It probably has little to no effect but I don't see any harm in using it.